crates/ra_hir/src/from_source.rs

```diff
@@ -129,6 +130,61 @@ impl FromSource for StructField {
     }
 }
 
+impl FromSource for AssocItem {
+    type Ast = ast::ImplItem;
+    fn from_source(db: &(impl DefDatabase + AstDatabase), src: Source<Self::Ast>) -> Option<Self> {
+        macro_rules! def {
+            ($kind:ident, $ast:ident) => {
+                $kind::from_source(db, Source { file_id: src.file_id, ast: $ast })
+                    .and_then(|it| Some(AssocItem::from(it)))
+            };
+        }
+
+        match src.ast {
+            ast::ImplItem::FnDef(f) => def!(Function, f),
```

Comment:
I think it's better to delegate to `Function::from_source` here, rather than re-implement